File: mailblastr-java/src/main/java/com/mailblastr/resources/Campaigns.java
package com.mailblastr.resources;

import com.mailblastr.ListParams;
import com.mailblastr.MailblastrResponse;
import com.mailblastr.http.ApiClient;
import com.mailblastr.requests.CreateCampaignRequest;
import com.mailblastr.requests.UpdateCampaignRequest;

import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Campaigns (DOMAIN-FIRST) — bulk sends to a domain's contact pool or one of
 * its segments. {@code domain} is REQUIRED on create.
 */
public final class Campaigns extends Resource {
    public Campaigns(ApiClient api) { super(api); }

    /** {@code POST /campaigns} */
    public MailblastrResponse create(CreateCampaignRequest request) {
        return api.request("POST", "/campaigns", request);
    }

    /** {@code GET /campaigns/:id} */
    public MailblastrResponse get(String id) {
        return api.request("GET", "/campaigns/" + enc(id));
    }

    /**
     * List campaigns. With no pagination params the route returns ALL of them
     * and {@code has_more:false}. {@code GET /campaigns}
     */
    public MailblastrResponse list() { return list(null); }

    public MailblastrResponse list(ListParams params) {
        return api.request("GET", "/campaigns" + paginate(params));
    }

    /** {@code PATCH /campaigns/:id} */
    public MailblastrResponse update(String id, UpdateCampaignRequest request) {
        return api.request("PATCH", "/campaigns/" + enc(id), request);
    }

    /** Send now. {@code POST /campaigns/:id/send} */
    public MailblastrResponse send(String id) {
        return api.request("POST", "/campaigns/" + enc(id) + "/send", Collections.emptyMap());
    }

    /** Schedule the send. {@code POST /campaigns/:id/send} with {@code { scheduled_at }}. */
    public MailblastrResponse send(String id, String scheduledAt) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("scheduled_at", scheduledAt);
        return api.request("POST", "/campaigns/" + enc(id) + "/send", body);
    }

    /**
     * Send (or schedule) with an IANA {@code schedule_timezone}, persisted
     * onto the campaign so daily batching evaluates batch-days in that zone.
     * Either argument may be {@code null} to omit it.
     * {@code POST /campaigns/:id/send}
     */
    public MailblastrResponse send(String id, String scheduledAt, String scheduleTimezone) {
        Map<String, Object> body = new LinkedHashMap<>();
        if (scheduledAt != null) body.put("scheduled_at", scheduledAt);
        if (scheduleTimezone != null) body.put("schedule_timezone", scheduleTimezone);
        return api.request("POST", "/campaigns/" + enc(id) + "/send", body);
    }

    /** Cancel a scheduled campaign (returns it to draft). {@code POST /campaigns/:id/cancel} */
    public MailblastrResponse cancel(String id) {
        return api.request("POST", "/campaigns/" + enc(id) + "/cancel");
    }

    /** Per-campaign analytics (counts, engagement rates, top links). {@code GET /campaigns/:id/stats} */
    public MailblastrResponse stats(String id) {
        return api.request("GET", "/campaigns/" + enc(id) + "/stats");
    }

    /**
     * Who opened / clicked / replied, per contact. Each of the three lists is
     * hard-capped at 500 rows and this route is not paginated.
     * {@code GET /campaigns/:id/engagement}
     */
    public MailblastrResponse engagement(String id) {
        return api.request("GET", "/campaigns/" + enc(id) + "/engagement");
    }

    /**
     * A/B winner evaluation for an A/B campaign. A campaign that is not an A/B
     * test answers {@code 422 validation_error}. Note the deliberately
     * camelCase {@code zScore} / {@code pValue} fields in the result.
     * {@code GET /campaigns/:id/ab}
     */
    public MailblastrResponse ab(String id) {
        return api.request("GET", "/campaigns/" + enc(id) + "/ab");
    }

    /** {@code DELETE /campaigns/:id} */
    public MailblastrResponse remove(String id) {
        return api.request("DELETE", "/campaigns/" + enc(id));
    }
}

File: mailblastr-rust/src/services/campaigns.rs
//! `mailblastr.campaigns` — DOMAIN-REQUIRED bulk sends to a domain's contact
//! pool (optionally scoped to a segment and/or gated by a topic), with A/B
//! tests, follow-ups, and recurrence. Types live in
//! [`super::campaign_types`].

use std::sync::Arc;

use reqwest::Method;
use serde_json::json;

use crate::client::{page_query, seg, Config};
use crate::services::campaign_types::{
    Campaign, CampaignAbResult, CampaignEngagement, CampaignListItem, CampaignStats,
    CreateCampaignOptions, SendCampaignOptions, UpdateCampaignOptions,
};
use crate::types::{IdResponse, ListResponse, PaginationParams, RemovedResponse, Result};

/// `mailblastr.campaigns`.
#[derive(Clone, Debug)]
pub struct CampaignsSvc {
    config: Arc<Config>,
}

impl CampaignsSvc {
    pub(crate) fn new(config: Arc<Config>) -> Self {
        Self { config }
    }

    /// Create a campaign (`domain` required). Returns `{ id }`.
    /// `POST /campaigns`
    pub async fn create(&self, options: CreateCampaignOptions) -> Result<IdResponse> {
        self.config
            .send(
                self.config
                    .request(Method::POST, "/campaigns")
                    .json(&options),
            )
            .await
    }

    /// Retrieve a campaign (includes statistics). `GET /campaigns/:id`
    pub async fn get(&self, campaign_id: &str) -> Result<Campaign> {
        let path = format!("/campaigns/{}", seg(campaign_id));
        self.config
            .send(self.config.request(Method::GET, &path))
            .await
    }

    /// List campaigns — trimmed [`CampaignListItem`] rows, with no bodies,
    /// follow-ups, recurrence or statistics. Use [`get`](Self::get) for those.
    /// `GET /campaigns`
    pub async fn list(
        &self,
        params: Option<PaginationParams>,
    ) -> Result<ListResponse<CampaignListItem>> {
        let req = self
            .config
            .request(Method::GET, "/campaigns")
            .query(&page_query(params.as_ref()));
        self.config.send(req).await
    }

    /// Update a draft campaign. Returns `{ id }`. `PATCH /campaigns/:id`
    pub async fn update(
        &self,
        campaign_id: &str,
        options: UpdateCampaignOptions,
    ) -> Result<IdResponse> {
        let path = format!("/campaigns/{}", seg(campaign_id));
        self.config
            .send(self.config.request(Method::PATCH, &path).json(&options))
            .await
    }

    /// Send now, or schedule with `scheduled_at`. For the optional
    /// `schedule_timezone` use [`send_with_options`](Self::send_with_options).
    /// `POST /campaigns/:id/send`
    pub async fn send(&self, campaign_id: &str, scheduled_at: Option<&str>) -> Result<IdResponse> {
        let path = format!("/campaigns/{}/send", seg(campaign_id));
        let body = match scheduled_at {
            Some(at) => json!({ "scheduled_at": at }),
            None => json!({}),
        };
        self.config
            .send(self.config.request(Method::POST, &path).json(&body))
            .await
    }

    /// Send now, or schedule with [`SendCampaignOptions`] — `scheduled_at`
    /// plus an optional `schedule_timezone` (IANA name, persisted onto the
    /// campaign so daily batching evaluates batch-days in that zone).
    /// `POST /campaigns/:id/send`
    pub async fn send_with_options(
        &self,
        campaign_id: &str,
        options: SendCampaignOptions,
    ) -> Result<IdResponse> {
        let path = format!("/campaigns/{}/send", seg(campaign_id));
        self.config
            .send(self.config.request(Method::POST, &path).json(&options))
            .await
    }

    /// Cancel a scheduled campaign (returns it to draft).
    /// `POST /campaigns/:id/cancel`
    pub async fn cancel(&self, campaign_id: &str) -> Result<Campaign> {
        let path = format!("/campaigns/{}/cancel", seg(campaign_id));
        self.config
            .send(self.config.request(Method::POST, &path))
            .await
    }

    /// Per-campaign analytics (counts, engagement rates, top links).
    /// `GET /campaigns/:id/stats`
    pub async fn stats(&self, campaign_id: &str) -> Result<CampaignStats> {
        let path = format!("/campaigns/{}/stats", seg(campaign_id));
        self.config
            .send(self.config.request(Method::GET, &path))
            .await
    }

    /// Per-recipient engagement (who opened / clicked / replied). Each list is
    /// capped at 500 rows; this route is not paginated.
    /// `GET /campaigns/:id/engagement`
    pub async fn engagement(&self, campaign_id: &str) -> Result<CampaignEngagement> {
        let path = format!("/campaigns/{}/engagement", seg(campaign_id));
        self.config
            .send(self.config.request(Method::GET, &path))
            .await
    }

    /// A/B winner evaluation for an A/B campaign. Answers
    /// `422 validation_error` when the campaign is not an A/B test.
    /// `GET /campaigns/:id/ab`
    pub async fn ab(&self, campaign_id: &str) -> Result<CampaignAbResult> {
        let path = format!("/campaigns/{}/ab", seg(campaign_id));
        self.config
            .send(self.config.request(Method::GET, &path))
            .await
    }

    /// Delete a campaign. `DELETE /campaigns/:id`
    pub async fn remove(&self, campaign_id: &str) -> Result<RemovedResponse> {
        let path = format!("/campaigns/{}", seg(campaign_id));
        self.config
            .send(self.config.request(Method::DELETE, &path))
            .await
    }
}
